rfcs/45-stageless.md

```diff
@@ -0,0 +1,996 @@
+To explain why, let's consider one common and significant case: `apply_system_buffers`.
+
+Suppose a user has written the following:
+
+```rust
+fn main() {
+    app
+        .add_system(apply_system_buffers)
+        .add_system(generates_commands.before(apply_system_buffers))
+        .add_system(relies_on_commands.after(apply_system_buffers))
+        .run()
+}
+```
+
+The user has added three systems and, naively, created dependencies using the `apply_system_buffers` type label.
+
+Since there's only one instance of `apply_system_buffers`, this app works like you'd expect.
+But now suppose that the user adds more anonymous instances of `apply_system_buffers` or imports plugin logic that does.
+
+Let's say the user imports a system set `X` from some plugin and `X` has another anonymous instance of `apply_system_buffers`.
+If the user schedules `X` to run before (or after) their logic, their program will panic over an unsolvable dependency graph.
+
+```rust
+fn main() {
+    app
+        .configure_set(X.before(generates_commands))
+        .add_system(apply_system_buffers)
+        .add_system(generates_commands.before(apply_system_buffers))
+        .add_system(relies_on_commands.after(apply_system_buffers))
+        // This will panic.
+        .run()
+}
+```
+
+What happened here?
+
+Well, the problem is a type-derived label will name *all* anonymous systems of that type.
+Therefore, `before(apply_system_buffers)` means "before all anonymous instances of `apply_system_buffers`" even though the user was most likely only thinking about the one they added themselves.
+Totally obvious, right? (/s)
+
+Their program panicked because it saw a contradiction: `generate_commands` had to run before all `apply_system_buffers` instances and run after the one inside `X` at the same time.
+The only (reliable) solution here is for the user to give their instance of `apply_system_buffers` a unique name and use that name for ordering.
+
+So while it's natural to think of systems as "sets of one", treating their type-derived labels as system set names is subtly different and doing so would virtually guarantee errors when multiple copies exist, so we just won't allow it.
+
+Now, we came up with three options to actually enforce that as an invariant:
+
+1. Require users to name duplicate systems.
+2. Leave duplicate systems completely anonymous. Only give the type-derived name to the first copy.
+3. Error when multiple anonymous copies of a system exist and their type-derived label is used for ordering.
+
+(1) ensures correctness and an easier debugging experience, but is too inconvenient.
+(2) drops the naming requirement without losing correctness, but makes things too unclear.
+(3) just makes doing the wrong thing an error.
+
+We like (3) because it doesn't put undue burden on the user or introduce unclear implicit behavior.
+Likewise, resolving the error is very simple: just name the systems.

```

Comment:
How will `chain![a, apply_system_buffers, b]` work in this context? I'm guessing that will be a common (desired) use case, but given that `chain!` is described as shorthand for before / after, this would result in conflicts with other `apply_system_buffers` instances.

I think I remember either @maniwani or @alice-i-cecile bringing up an "ascending unique system id" concept on discord (or alternatively, we could bring back `System::id()`)? If we add that, then `chain![a, apply_system_buffers, b]` could internally use those for ordering, removing the need for internally using `a.after(apply_system_buffers)`, which would produce a warning if there is more than one `apply_system_buffers`, as outlined in this section in the RFC. That would cut down on the number of cases where manually labeling `apply_system_buffers` (or other systems) is necessary.

Reply:
Actually maybe it just makes sense to use the `NodeId` as defined in this RFC?

Reply:
We could then remove the comment from this part of the RFC: https://github.com/bevyengine/rfcs/pull/45/files#diff-d319c5378a207f7eae8b553b1df6b026d7bfc885b016f776a64b48ed86935d5fR274

Reply:
My gut instinct here is that most of the cases where you're using non-unique systems (especially in a chain) fall into the "at-least-once-seperation" patterns, and a truly good solution requires a fix for that (or you end up inserting tons of command flushes or other costly systems mostly implicitly).

That said, I think that your `NodeId` solution will work to avoid the immediate pain. Let me chew on that.

Reply:
I've incorporated your ideas in https://github.com/bevyengine/rfcs/pull/45/commits/560b1f11de7da83592f457e2ca90bf3c72713471

Reply:
Yes, using `NodeId` was how I intended to create these edges.